ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/BeaconStateUtil.java

```diff
@@ -668,165 +437,21 @@ public static Bytes32 get_randao_mix(BeaconState state, UnsignedLong epoch) {
-   *
-   * <p>Utilizes 'swap or not' shuffling found in
-   * https://link.springer.com/content/pdf/10.1007%2F978-3-642-32009-5_1.pdf. See the 'generalized
-   * domain' algorithm on page 3.
-   *
-   * @param index The index in the permuatation we wish to get the value of.
-   * @param list_size The size of the list from which the element is taken.
-   * @param seed Initial seed value used for randomization.
-   * @return The index from the original list that is now at position `index`
-   * @see <a
-   *     href="https://github.com/ethereum/eth2.0-specs/blob/v0.4.0/specs/core/0_beacon-chain.md#get_permuted_index">get_permuted_index
-   *     - Spec v0.4</a>
-   */
-  @VisibleForTesting
-  public static int get_permuted_index(int index, int list_size, Bytes32 seed) {
-    checkArgument(index < list_size, "get_permuted_index(): index greater than list size");
-
-    // The spec says that we should handle up to 2^40 validators, but we can't do this,
-    // so we just fall back to int (2^31 validators).
-    // checkArgument(list_size <= 1099511627776L); // 2^40
-
-    /*
-     * In the following, great care is needed around signed and unsigned values.
-     * Note that the % (modulo) operator in Java behaves differently from the
-     * modulo operator in python:
-     *   Python -1 % 13 = 12
-     *   Java   -1 % 13 = -1
-     *
-     * Using UnsignedLong doesn't help us as some quantities can legitimately be negative.
-     */
-
-    int indexRet = index;
-    byte[] powerOfTwoNumbers = {1, 2, 4, 8, 16, 32, 64, (byte) 128};
-
-    for (int round = 0; round < SHUFFLE_ROUND_COUNT; round++) {
-
-      Bytes roundAsByte = Bytes.of((byte) round);
-
-      // This needs to be unsigned modulo.
-      int pivot =
-          toIntExact(
-              Long.remainderUnsigned(
-                  bytes_to_int(Hash.keccak256(Bytes.wrap(seed, roundAsByte)).slice(0, 8)),
-                  list_size));
-      int flip = (pivot - indexRet) % list_size;
-      if (flip < 0) {
-        // Account for flip being negative
-        flip += list_size;
-      }
-
-      int position = (indexRet < flip) ? flip : indexRet;
-
-      Bytes positionDiv256 = int_to_bytes(position / 256, 4);
-      Bytes source = Hash.keccak256(Bytes.wrap(seed, roundAsByte, positionDiv256));
-
-      // The byte type is signed in Java, but the right shift should be fine as we just use bit 0.
-      // But we can't use % in the normal way because of signedness, so we `& 1` instead.
-      byte theByte = source.get(position % 256 / 8);
-      byte theMask = powerOfTwoNumbers[position % 8];
-      if ((theByte & theMask) != 0) {
-        indexRet = flip;
-      }
-    }
-
-    return indexRet;
+                    UnsignedLong.valueOf(Math.floorDiv(SHARD_COUNT, SLOTS_PER_EPOCH)),
+                    UnsignedLong.valueOf(Math.floorDiv(active_validator_indices.size(), Math.floorDiv(SLOTS_PER_EPOCH, TARGET_COMMITTEE_SIZE)))
+            )
+    ).times(UnsignedLong.valueOf(SLOTS_PER_EPOCH));
```

Comment:
You don't need to use Math.floorDiv when working with UnsignedLong. The only time floorDiv is different that integer division with truncation is when the integer is negative. This can't happen with UnsignedLongs:
https://docs.oracle.com/javase/8/docs/api/java/lang/Math.html#floorDiv-long-long-